PLNSRVCE-658: fix deployment get/create, failed attempt to fix service get/create

gabemontero opened the pull request:
So with an assist from @Mo3m3n (I think I have that git ID correct, apologies if it was not), we were able to adjust the way jvm-build-service made its controller-runtime manager "kcp aware", which ended up fixing the HTTP URL used on the POST to create the deployments for the artifact cache and localstack.

With log level 6 still, you will see for example:

```
I0926 18:26:54.317774       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/apis/apps/v1/deployments?limit=500&resourceVersion=0 200 OK in 4 milliseconds
I0926 18:26:54.323793       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/apis/apps/v1/deployments?allowWatchBookmarks=true&resourceVersion=2590&timeoutSeconds=590&watch=true 200 OK in 5 milliseconds
I0926 18:26:54.413683       1 shared_informer.go:285] caches populated
1.6642168144138787e+09	INFO	userconfig	GetImageName controller image quay.io/gabemontero/hacbs-jvm-controller:dev	{"request": "jvm-build-service/jvm-build-config", "cluster": "root:default:pipeline-service-compute"}
1.664216814413896e+09	INFO	userconfig	GetImageName using quay.io/gabemontero/hacbs-jvm-cache:dev for hacbs-jvm-cache	{"request": "jvm-build-service/jvm-build-config", "cluster": "root:default:pipeline-service-compute"}
I0926 18:26:54.425015       1 round_trippers.go:553] POST https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/root:default:pipeline-service-compute/apis/apps/v1/namespaces/jvm-build-service/deployments 201 Created in 10 milliseconds
```

I am even able to get/create some core k8s types, like serviceaccounts:

```
I0926 18:26:54.198470       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/api/v1/serviceaccounts?limit=500&resourceVersion=0 200 OK in 4 milliseconds
I0926 18:26:54.202824       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/api/v1/serviceaccounts?allowWatchBookmarks=true&resourceVersion=2590&timeoutSeconds=410&watch=true 200 OK in 3 milliseconds
I0926 18:26:54.294229       1 shared_informer.go:285] caches populated
I0926 18:26:54.301770       1 round_trippers.go:553] POST https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/root:default:pipeline-service-compute/api/v1/namespaces/jvm-build-service/serviceaccounts 201 Created in 7 milliseconds
```

However, inexplicably to me secrets and services has proven unsuccessful. I've hence commented out the controllers secret actions for now to simplify debug and focus on serviceaccounts.  The log level 6 output produced first 

```
1.6642168141801999e+09	INFO	userconfig	GGM2 before get of service host https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice apipath  req cluster root:default:pipeline-service-computes	{"request": "jvm-build-service/jvm-build-config", "cluster": "root:default:pipeline-service-compute"}
I0926 18:26:54.185917       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/api?timeout=32s 200 OK in 5 milliseconds
I0926 18:26:54.189476       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/apis?timeout=32s 200 OK in 3 milliseconds
I0926 18:26:54.193297       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/apis/tekton.dev/v1beta1?timeout=32s 200 OK in 3 milliseconds
I0926 18:26:54.193308       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/api/v1?timeout=32s 200 OK in 3 milliseconds
I0926 18:26:54.193412       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/apis/jvmbuildservice.io/v1alpha1?timeout=32s 200 OK in 3 milliseconds
I0926 18:26:54.193438       1 round_trippers.go:553] GET https://ckcp-ckcp.apps.gmontero410.devcluster.openshift.com:443/services/apiexport/root:default:pipeline-service-compute/jvmbuildservice/clusters/%2A/apis/apps/v1?timeout=32s 200 OK in 3 milliseconds
1.6642168141937797e+09	INFO	userconfig	GGM2 after get of service error &meta.NoKindMatchError{GroupKind:schema.GroupKind{Group:"", Kind:"Service"}, SearchedVersions:[]string{"v1"}}	{"request": "jvm-build-service/jvm-build-config", "cluster": "root:default:pipeline-service-compute"}
```

Followed soon after by
```
1.6642168144448457e+09	ERROR	Reconciler error	{"controller": "userconfig", "controllerGroup": "jvmbuildservice.io", "controllerKind": "UserConfig", "userConfig": {"name":"jvm-build-config","namespace":"jvm-build-service"}, "namespace": "jvm-build-service", "name": "jvm-build-config", "reconcileID": "552fd78c-7ba6-4837-a4df-b0f33ac2dd72", "error": "no matches for kind \"Service\" in version \"v1\""}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234
```

gabemontero commented:
apibinding:

```
gmontero ~/go/src/github.com/redhat-appstudio/jvm-build-service  (fix-create-client)$ KUBECONFIG=../../openshift-pipelines/pipeline-service/ckcp/work/admin.kubeconfig oc get apibindings jvmbuildservice-binding -o yaml 
apiVersion: apis.kcp.dev/v1alpha1
kind: APIBinding
metadata:
  annotations:
    kcp.dev/cluster: root:default:pipeline-service-compute
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apis.kcp.dev/v1alpha1","kind":"APIBinding","metadata":{"annotations":{},"name":"jvmbuildservice-binding"},"spec":{"permissionClaims":[{"group":"","resource":"secrets","state":"Accepted"},{"group":"","resource":"serviceaccounts","state":"Accepted"},{"group":"","resource":"services","state":"Accepted"},{"group":"apps","identityHash":"084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50","resource":"deployments","state":"Accepted"},{"group":"tekton.dev","identityHash":"084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50","resource":"pipelineruns","state":"Accepted"}],"reference":{"workspace":{"exportName":"jvmbuildservice","path":"root:default:pipeline-service-compute"}}}}
  creationTimestamp: "2022-09-26T18:24:57Z"
  finalizers:
  - apis.kcp.dev/apibinding-finalizer
  generation: 1
  labels:
    claimed.internal.apis.kcp.dev/7U0v4qf3SKgUyVEDLlgSlI4d1xpK3sfm2955c: LstvmbbzVDDOn90ZbhzSQO5U3DMCf88h1pZ
    internal.apis.kcp.dev/export: 7U0v4qf3SKgUyVEDLlgSlI4d1xpK3sfm2955c
  name: jvmbuildservice-binding
  resourceVersion: "2581"
  uid: 023e8656-29d5-4c00-a88e-57bb2ac82dc7
spec:
  permissionClaims:
  - resource: secrets
    state: Accepted
  - resource: serviceaccounts
    state: Accepted
  - resource: services
    state: Accepted
  - group: apps
    identityHash: 084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50
    resource: deployments
    state: Accepted
  - group: tekton.dev
    identityHash: 084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50
    resource: pipelineruns
    state: Accepted
  reference:
    workspace:
      exportName: jvmbuildservice
      path: root:default:pipeline-service-compute
status:
  appliedPermissionClaims:
  - group: tekton.dev
    identityHash: 084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50
    resource: pipelineruns
  - resource: secrets
  - resource: serviceaccounts
  - resource: services
  - group: apps
    identityHash: 084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50
    resource: deployments
  boundExport:
    workspace:
      exportName: jvmbuildservice
      path: root:default:pipeline-service-compute
  boundResources:
  - group: jvmbuildservice.io
    resource: artifactbuilds
    schema:
      UID: ba439780-300a-4436-8f20-3e6136a7b3a8
      identityHash: 532d7412903610c2912c682fd81c8a71aef96034a435ccfd9be3225040fb7968
      name: v202209260410.artifactbuilds.jvmbuildservice.io
    storageVersions:
    - v1alpha1
  - group: jvmbuildservice.io
    resource: dependencybuilds
    schema:
      UID: dbd6d6b0-5421-48ec-b4e8-745aa30a5955
      identityHash: 532d7412903610c2912c682fd81c8a71aef96034a435ccfd9be3225040fb7968
      name: v202209260410.dependencybuilds.jvmbuildservice.io
    storageVersions:
    - v1alpha1
  - group: jvmbuildservice.io
    resource: rebuiltartifacts
    schema:
      UID: e5a984de-9938-49a3-ac78-8fe07e1078b3
      identityHash: 532d7412903610c2912c682fd81c8a71aef96034a435ccfd9be3225040fb7968
      name: v202209260410.rebuiltartifacts.jvmbuildservice.io
    storageVersions:
    - v1alpha1
  - group: jvmbuildservice.io
    resource: systemconfigs
    schema:
      UID: 6ac8d0cb-1774-4d53-a7a5-6b75cbd9ea44
      identityHash: 532d7412903610c2912c682fd81c8a71aef96034a435ccfd9be3225040fb7968
      name: v202209260410.systemconfigs.jvmbuildservice.io
    storageVersions:
    - v1alpha1
  - group: jvmbuildservice.io
    resource: tektonwrappers
    schema:
      UID: 9ac03be3-2f21-40bb-ad4e-e4fc50ba6fe4
      identityHash: 532d7412903610c2912c682fd81c8a71aef96034a435ccfd9be3225040fb7968
      name: v202209260410.tektonwrappers.jvmbuildservice.io
    storageVersions:
    - v1alpha1
  - group: jvmbuildservice.io
    resource: userconfigs
    schema:
      UID: c729ad14-cb20-474e-a323-1d65c3a5290b
      identityHash: 532d7412903610c2912c682fd81c8a71aef96034a435ccfd9be3225040fb7968
      name: v202209260410.userconfigs.jvmbuildservice.io
    storageVersions:
    - v1alpha1
  conditions:
  - lastTransitionTime: "2022-09-26T18:24:58Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2022-09-26T18:24:58Z"
    status: "True"
    type: APIExportValid
  - lastTransitionTime: "2022-09-26T18:24:58Z"
    status: "True"
    type: BindingUpToDate
  - lastTransitionTime: "2022-09-26T18:24:58Z"
    status: "True"
    type: InitialBindingCompleted
  - lastTransitionTime: "2022-09-26T18:24:58Z"
    status: "True"
    type: PermissionClaimsApplied
  - lastTransitionTime: "2022-09-26T18:24:58Z"
    status: "True"
    type: PermissionClaimsValid
  exportPermissionClaims:
  - resource: secrets
  - resource: serviceaccounts
  - resource: services
  - group: apps
    identityHash: 084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50
    resource: deployments
  - group: tekton.dev
    identityHash: 084e59e91659255ee5a002d5acafe69a42cb6e52035506bfa3619d0e8f17ef50
    resource: pipelineruns
  phase: Bound
gmontero ~/go/src/github.com/redhat-appstudio/jvm-build-service  (fix-create-client)$ 
```

gabemontero commented:
I was also able to verify @stuartwdouglas that I could create an artfiactbuild from the controller (I did it off a `userconfig` event) and
- the artifactbuild reconciler got the event
- the artifactbuild could be seen from the kcp enabled CLI / logged in the kcp context

so "meets min" there

Not suprisingly, you cannot see the artifactbuild from the workload cluster context.  Lines up with all the info around sync that I have heard. The CRDs are not even available on the workload cluster.  Presumably, pipelineruns will be though.  We'll know for sure once the remaining barriers to bring up are addressed.

gabemontero commented:
ok base line e2e is fine

I am curious how the move from a pvc to an ephemeral volume affects the artifact cache with a bigger build like apicurio

/test jvm-build-service-in-repo-apicurio-e2e

Mo3m3n commented on deploy/kcp/apiexport_jvmbuildservice.yaml:
```diff
@@ -10,15 +10,17 @@ spec:
-    resource: "configmaps"
+    resource: "serviceaccounts"
+  - group: ""
+    resource: "services"
```
`services` resource is not a kcp built-in type so you need to claim it via the identityHash referenced in kubernetes APIExport. Just add the same hash as in deployments and pipelineruns and also update the APIBinding accordingly and that probably should fix the get/create issue for services

gabemontero commented:
well, artifact gathering died, so we didn't get our typical set of logs 

the artifact cache ran for a bit, dowloaded a bunch of stuff, and at one point got

```
2022-09-26 21:33:54,304 ERROR [com.red.hac.art.ser.RepositoryCache$DownloadingFile] (executor-thread-43) Failed to download artifact /cache/jsweet/original/io/undertow/undertow-servlet/1.4.0.CR5-SNAPSHOT/undertow-servlet-1.4.0.CR5-SNAPSHOT.pom from MavenClient{name='jsweet', uri=https://repository.jsweet.org/artifactory/libs-release-local}: java.lang.RuntimeException: java.lang.RuntimeException: Unexpected status code: 409
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:116)
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.getArtifactFile(MavenClient.java:57)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.lambda$getArtifactFile$0(RepositoryCache.java:70)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache$DownloadingFile.download(RepositoryCache.java:231)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.handleFile(RepositoryCache.java:107)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.getArtifactFile(RepositoryCache.java:69)
	at com.redhat.hacbs.artifactcache.services.CacheFacade.getArtifactFile(CacheFacade.java:54)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource.filterNewerVersions(CacheMavenResource.java:121)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource.get(CacheMavenResource.java:80)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource$quarkusrestinvoker$get_b06f4dda75756eadf80a70f5b08a2d5bf4ce49bb.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:108)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:140)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:557)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: Unexpected status code: 409
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:98)
	... 19 more
```

I don't have a sense if moving from a PV to an EV would have any bearing on that.

Let's try again

/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
hmmm the installing of kcp tooling is failing ... maybe we have to bump some versions ?  going to look at some of the other app studio repos

gabemontero commented:
tekton failed to install

/retest

that said, I'm seeing weird stuff locally since the removal of localstack; hopefully, we get a clean run here to cross reference

gabemontero commented on deploy/base/config.yaml:
```diff
@@ -18,7 +18,6 @@ apiVersion: jvmbuildservice.io/v1alpha1
 kind: UserConfig
 metadata:
   name: jvm-build-config
-  namespace: test-jvm-namespace
```
@stuartwdouglas - I think you added this back to my last kcp PR.

For the dev flow, should we not have already switched to the test-jvm-namespace per https://github.com/redhat-appstudio/jvm-build-service/blob/main/deploy/base-development.sh#L13 before this gets applied at https://github.com/redhat-appstudio/jvm-build-service/blob/main/deploy/base-development.sh#L19 ?

Not specifying the namespace here provides some flexibility to use this config in our scenarios besides our classic dev flow like some of the kcp testing I am doing.

stuartwdouglas replied:
I was seeing a failure without it, I am not sure why. I did not actually mean to commit it until I had investigated exactly what was going on.

gabemontero replied:
ok good to know

I'll try the full openshift dev flow today (vs. the openshift ci flow I usually do) and see if I can reproduced it today.  Maybe it was a minikube only thing?

But let's plan on leaving the namespace setting removed here.  At worst, it is an easy work around for whatever flow needed it.

stuartwdouglas replied:
Without it I see:

```
error: no matches for Id jvmbuildservice.io_v1alpha1_UserConfig|test-jvm-namespace|jvm-build-config; failed to find unique target for patch jvmbuildservice.io_v1alpha1_UserConfig|jvm-build-config
```

This is on normal openshift with a new clean namespace.

stuartwdouglas replied:
Ah, you just need to delete the namespace from dev-template/config.yaml as well.

gabemontero replied:
ok will do that after my apicurio job finishes

thanks

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

stuartwdouglas commented:
/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
ahh ... also found 

```
1.664340409839929e+09	ERROR	Reconciler error	{"controller": "userconfig", "controllerGroup": "jvmbuildservice.io", "controllerKind": "UserConfig", "userConfig": {"name":"jvm-build-config","namespace":"jvm-build-service-test-namespace-4ngr5"}, "namespace": "jvm-build-service-test-namespace-4ngr5", "name": "jvm-build-config", "reconcileID": "bff933c1-05e6-4d9f-85f1-b71def58190c", "error": "secrets \"jvm-build-secrets\" not found"}
        sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273
        sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234
```

that is definitely related to my change :-)

gabemontero commented:
> > ahh ... also found
> > ```
> > 1.664340409839929e+09	ERROR	Reconciler error	{"controller": "userconfig", "controllerGroup": "jvmbuildservice.io", "controllerKind": "UserConfig", "userConfig": {"name":"jvm-build-config","namespace":"jvm-build-service-test-namespace-4ngr5"}, "namespace": "jvm-build-service-test-namespace-4ngr5", "name": "jvm-build-config", "reconcileID": "bff933c1-05e6-4d9f-85f1-b71def58190c", "error": "secrets \"jvm-build-secrets\" not found"}
> >         sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
> >         	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273
> >         sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
> >         	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > that is definitely related to my change :-)
> 
> false alarm - the secret shows up as expected the the artifact cache deployment is still created
> 
> `I0928 04:46:50.330581 1 round_trippers.go:553] POST https://172.30.0.1:443/apis/apps/v1/namespaces/jvm-build-service-test-namespace-4ngr5/deployments 201 Created in 18 milliseconds`

well, not only did I not get that formatting error seen in CI, today I am not seeing any of the weirdness I saw yesterday per https://coreos.slack.com/archives/C03KN2B47GU/p1664311929444089

I'm going to ignore that irritating mystery for now and am tracking the artifact cache's health while the service registry run progresses in my local setup.

gabemontero commented:
fortunately @stuartwdouglas remembered what caused https://github.com/redhat-appstudio/jvm-build-service/pull/270#issuecomment-1260892630

see https://github.com/redhat-appstudio/jvm-build-service/pull/270/commits/a9bd591d47709800f7f5226bd54a661ca0d981b3

gabemontero commented:
yeah none of those S3 exceptions show up in the artifact cache in the PR's current e2e run  @stuartwdouglas 

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-e2e/1575246844068368384/artifacts/jvm-build-service-in-repo-e2e/gather-extra/artifacts/pods/jvm-build-service-test-namespace-rdsp2_jvm-build-workspace-artifact-cache-76c9d75b99-mvnjw_jvm-build-workspace-artifact-cache.log

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

stuartwdouglas commented:
/retest

gabemontero commented:
so the initial apicurio build was just stuck in the clone step

```
    types.go:25: time: 2022-09-29 06:35:06.989709438 +0000 UTC m=+633.621557971: pod logs for container step-clone in pod service-registry-build-x4x4b-fetch-repository-pod:  + '[' false = true ']'
        + '[' false = true ']'
        + CHECKOUT_DIR=/workspace/output/
        + '[' true = true ']'
        + cleandir
        + '[' -d /workspace/output/ ']'
        + rm -rf /workspace/output//lost+found
        + rm -rf '/workspace/output//.[!.]*'
        + rm -rf '/workspace/output//..?*'
        + test -z ''
        + test -z ''
        + test -z ''
        + /ko-app/git-init -url=https://github.com/Apicurio/apicurio-registry.git -revision=2.2.4.Final -refspec= -path=/workspace/output/ -sslVerify=true -submodules=true -depth=1 -sparseCheckoutDirectories=
        {"level":"info","ts":1664432745.4581563,"caller":"git/git.go:169","msg":"Successfully cloned https://github.com/Apicurio/apicurio-registry.git @ d8dc51fef36dc3b2a6e12a9362c99d60a8cd7e09 (grafted, HEAD) in path /workspace/output/"}
        {"level":"info","ts":1664432745.485094,"caller":"git/git.go:207","msg":"Successfully initialized and updated submodules in path /workspace/output/"}
        + cd /workspace/output/
        ++ git rev-parse HEAD
        + RESULT_SHA=d8dc51fef36dc3b2a6e12a9362c99d60a8cd7e09
        + EXIT_CODE=0
        + '[' 0 '!=' 0 ']'
        + printf %!s(MISSING) d8dc51fef36dc3b2a6e12a9362c99d60a8cd7e09
        + printf %!s(MISSING) https://github.com/Apicurio/apicurio-registry.git
    types.go:25: time: 2022-09-29 06:35:06.989764209 +0000 UTC m=+633.621612742: dumpPods looking at pod service-registry-build-x4x4b-maven-run-pod in phase Pending
    types.go:25: time: 2022-09-29 06:35:07.055811608 +0000 UTC m=+633.687660151: error getting pod logs for container step-mvn-settings: container "step-mvn-settings" in pod "service-registry-build-x4x4b-maven-run-pod" is waiting to start: PodInitializing
    types.go:25: time: 2022-09-29 06:35:07.121672384 +0000 UTC m=+633.753520926: error getting pod logs for container step-mvn-goals: container "step-mvn-goals" in pod "service-registry-build-x4x4b-maven-run-pod" is waiting to start: PodInitializing
    types.go:25: time: 2022-09-29 06:35:07.187716902 +0000 UTC m=+633.819565445: error getting pod logs for container step-analyse-dependencies: container "step-analyse-dependencies" in pod "service-registry-build-x4x4b-maven-run-pod" is waiting to start: PodInitializing
```

now, I do not see any bad events around unhealthy pods, constrained pods etc.

However, after closer examination of all the events, including normal ones, it looks like the fetch repo step restarted once or twice, eventually completed, and the build started, but by then, the test had given up.

I also noticed that the git clone step uses a PVC (not surprising, but had not noticed that before).  I believe that is OK, based on what I've seen with KCP testing before, and where tekton objs sit in the compute cluster vs. kcp cluster equation.

Pretty sure the default 10 minute timeout for the initial build is too close to call for the service registry build.  I'm going to bump that, and try again.

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
well, it got past the clone this time
unfortunately, it did not complete in time, but some failed dependency builds logs were captured in the test artifacts, and in the must gather artifacts at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1575488288238604288/artifacts/jvm-build-service-in-repo-apicurio-e2e/gather-extra/artifacts/pods/ you see some successful ones as well

the artifact cache at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1575488288238604288/artifacts/jvm-build-service-in-repo-apicurio-e2e/gather-extra/artifacts/pods/jvm-build-service-test-namespace-hv887_jvm-build-workspace-artifact-cache-78b59d68bd-zp6xx_jvm-build-workspace-artifact-cache.log looks clean

the controller log looks clean too at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1575488288238604288/artifacts/jvm-build-service-in-repo-apicurio-e2e/gather-extra/artifacts/pods/jvm-build-service_hacbs-jvm-operator-5d8b78fd79-cdkc7_hacbs-jvm-operator.log

and I cannot find any bad events from the raw events dump

going to push an update for the non-caching client, as well as clean up debug and squash commits, and we'll try one more time

based on the results, and my still unable to run things locally, I may see if @stuartwdouglas can run a service registry build with this branch

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

stuartwdouglas commented:
/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
we actually got logs with https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1575552461958025216 @stuartwdouglas 

some interesting results .... still trying to determine if the artifact cache using an ephemeral volume / being memory constrained has any bearing on them

the last dependencybuild count is 

`    types.go:25: time: 2022-09-29 21:20:38.925692937 +0000 UTC m=+7178.977179860: dependencybuild created count: 116 complete count: 52, failed count: 20, contaminated count: 0`

the controller logs at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1575552461958025216/artifacts/jvm-build-service-in-repo-apicurio-e2e/gather-extra/artifacts/pods/jvm-build-service_hacbs-jvm-operator-66b6659f46-xvcm2_hacbs-jvm-operator.log look OK to me

in the artifact cache, no S3 exception, and "generally" clean.  I do see 
- sporadic `INFO  [com.red.hac.art.res.CacheMavenResource] (executor-thread-11) Failed to get artifact ...`
- given it is an INFO, I'm assuming it is sporadic
- there is one error stack trace:
```
2022-09-29 19:33:09,060 SEVERE [com.red.hac.cla.tra.ClassFileTracker] (executor-thread-23) Failed to add tracking data to class: module-info.class: java.lang.ArrayIndexOutOfBoundsException: Index 29 out of bounds for length 9
	at org.objectweb.asm.ClassReader.readUtf(ClassReader.java:3669)
	at org.objectweb.asm.ClassReader.readUTF8(ClassReader.java:3656)
	at org.objectweb.asm.ClassReader.readStringish(ClassReader.java:3726)
	at org.objectweb.asm.ClassReader.readModule(ClassReader.java:3756)
	at org.objectweb.asm.ClassReader.readModuleAttributes(ClassReader.java:777)
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:580)
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
	at com.redhat.hacbs.classfile.tracker.ClassFileTracker.addTrackingDataToClass(ClassFileTracker.java:29)
	at com.redhat.hacbs.classfile.tracker.ClassFileTracker.addTrackingDataToJar(ClassFileTracker.java:68)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.handleDownloadedFile(RepositoryCache.java:158)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache$DownloadingFile.download(RepositoryCache.java:309)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.handleFile(RepositoryCache.java:107)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.getArtifactFile(RepositoryCache.java:69)
	at com.redhat.hacbs.artifactcache.services.CacheFacade.getArtifactFile(CacheFacade.java:54)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource.get(CacheMavenResource.java:48)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource$quarkusrestinvoker$get_8d5780d6a5750c6cc9c134b1a07b3dc9855eee94.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:108)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:140)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:557)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)

```

Some of the dependencybuild failures are not familiar to me.  Maybe you or @dwalluck could look at them and decipher if they are in anyway caused by a constrained cache / system?

I'm still looking at data, but if I had to guess, *perhaps* things are constrained either because
- there is so much memory assigned to the cache, it slows down other processing
- or I need to give more memory to the cache. I gave it 8Gi, but the volume has a max of 10Gi.  Maybe we are bumping up against memory limits.  I could try going to 11 Gi (10Gi for the disk, 1 Gi for the original mem default limit).

WDYT @stuartwdouglas ?

gabemontero commented:
yeah none of the nodes have memory pressure

If I had to guess, it is just "slower" with the current mem limit setting and no PV

Unless you all see something weird in the db failure logs

stuartwdouglas commented:
I will go through the logs a bit later, I am trying to debug why my PR is failing ATM, I think there is an actual issue, even though it  runs fine locally.

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
another instance of where it ran, we see it gather some failed db's, but the entire log collection failed ^^

stuartwdouglas commented:
I think it is running out of memory gathering the logs. We should try streaming from the log collector straight to the file, rather than reading the full logs into memory.

gabemontero commented:
> I think it is running out of memory gathering the logs. We should try streaming from the log collector straight to the file, rather than reading the full logs into memory.

oooh that make sense 

I'll work on that today, building off of what you do for the ab/db artifact gathering

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
ok with my local builds working again I have an apicurio job running in parallel with the CI job here and I have bash scripts running that loops on monitoring both node cpu/memory usage and the artifact cache's cpu/memory usage on my 2xlarge 3 nodes at 32 Gib cluster.

A quick sample:

```
NAME                                                  CPU(cores)   MEMORY(bytes)   
jvm-build-workspace-artifact-cache-5876d8c487-kt5pr   0m           1513Mi          
NAME                                         CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
ip-10-0-140-141.us-west-1.compute.internal   192m         2%     7860Mi          25%       
ip-10-0-152-199.us-west-1.compute.internal   439m         12%    8107Mi          55%       
ip-10-0-167-191.us-west-1.compute.internal   324m         4%     13519Mi         44%       
ip-10-0-180-49.us-west-1.compute.internal    554m         15%    7425Mi          50%       
ip-10-0-204-29.us-west-1.compute.internal    352m         4%     14244Mi         46%       
ip-10-0-232-71.us-west-1.compute.internal    731m         20%    10233Mi         69%       
---
```

I'll see how far it gets in the current 3 hours we've currently allocated for the apicurio test and see where things land.

At the moment, it doesn't look like we need more than 2 to 3 Gi for the artifact cache when using ephemeral volumes.  This does confirm that the earlier instance of the cache falling over and restarting I saw at the default 1Gi make sense since if has been sitting at 1.5 Gi for a bit.

@stuartwdouglas FYI

the command to monitor like this is `oc adm top ...`

stuartwdouglas commented:
Not much detail in the failures, it seems to have timed out without attempting to generate the report, but there are lots of builds that have happened in gather-extra.

gabemontero commented:
So when the test timed out for me locally, because of the time out with the test framework, none of our failure gathering kicked in.
However, I of course still had access to the cluster @stuartwdouglas 

Turns the current set of active pipeline runs had been active for an hour.

Too long, right.

I tailed their live logs, and they all were stuck trying to access the cache.  The same pattern:
```
[task : build] [INFO] Downloading from mirror.default: http://jvm-build-workspace-artifact-cache.jvm-build-service-test-namespace-6gssb.svc.cluster.local/v1/cache/default/1644883013000/io/smallrye/smallrye-parent/34/smallrye-parent-34.pom
[task : build] [INFO] I/O exception (org.apache.http.NoHttpResponseException) caught when processing request to {}->http://jvm-build-workspace-artifact-cache.jvm-build-service-test-namespace-6gssb.svc.cluster.local:80: The target server failed to respond
[task : build] [INFO] Retrying request to {}->http://jvm-build-workspace-artifact-cache.jvm-build-service-test-namespace-6gssb.svc.cluster.local:80
```

and then nothing.  But the pipelinerun is still running.

Over on the artifact cache side, I see no obvious signs of distress.  The pod memory in constant, little cpu usage.  The node is not under pressure.  Liveness probes OK.

In the middle of the artifact build log, there is this one exception:
```
2022-09-30 20:06:41,201 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-25) HTTP Request to /v1/cache/default/1568202546000/org/glassfish/copyright/glassfish-copyright-maven-plugin/maven-metadata.xml.sha1 failed, error id: f104c2fb-01a0-44b0-aa76-f9d8a8241577-1: java.lang.RuntimeException: java.lang.RuntimeException: java.net.SocketException: Connection reset
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:116)
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.getMetadataFile(MavenClient.java:123)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.getMetadataFile(RepositoryCache.java:74)
	at com.redhat.hacbs.artifactcache.services.CacheFacade.getMetadataFiles(CacheFacade.java:78)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource.get(CacheMavenResource.java:71)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource$quarkusrestinvoker$get_b06f4dda75756eadf80a70f5b08a2d5bf4ce49bb.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:108)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:140)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:557)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: java.net.SocketException: Connection reset
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:90)
	... 15 more
Caused by: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:323)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:350)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:803)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:966)
	at java.base/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:478)
	at java.base/sun.security.ssl.SSLSocketInputRecord.readHeader(SSLSocketInputRecord.java:472)
	at java.base/sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(SSLSocketInputRecord.java:70)
	at java.base/sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1455)
	at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1059)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:280)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
	at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:157)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:88)
	... 15 more

2022-09-30 20:06:41,202 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-25) Request failed: java.lang.RuntimeException: java.lang.RuntimeException: java.net.SocketException: Connection reset
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:116)
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.getMetadataFile(MavenClient.java:123)
	at com.redhat.hacbs.artifactcache.services.RepositoryCache.getMetadataFile(RepositoryCache.java:74)
	at com.redhat.hacbs.artifactcache.services.CacheFacade.getMetadataFiles(CacheFacade.java:78)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource.get(CacheMavenResource.java:71)
	at com.redhat.hacbs.artifactcache.resources.CacheMavenResource$quarkusrestinvoker$get_b06f4dda75756eadf80a70f5b08a2d5bf4ce49bb.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:108)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:140)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:557)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: java.net.SocketException: Connection reset
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:90)
	... 15 more
Caused by: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:323)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:350)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:803)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:966)
	at java.base/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:478)
	at java.base/sun.security.ssl.SSLSocketInputRecord.readHeader(SSLSocketInputRecord.java:472)
	at java.base/sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(SSLSocketInputRecord.java:70)
	at java.base/sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1455)
	at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1059)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:280)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
	at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:157)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at com.redhat.hacbs.artifactcache.services.client.maven.MavenClient.downloadMavenFile(MavenClient.java:88)
	... 15 more

```

But then a bunch of normal looking logs after that.

Any chance that retry logic in the pipeline runs has a bug and its hung somehow @stuartwdouglas ?

I'm going to work on augmenting the e2e test to look how long the active pipeline runs have been running, and if it goes past say 30 or 45 minutes, gather those logs and see if we see the same thing ^^

I'm also going to put a time limit on our last t.Run so that we make sure we gather our failed logs before the testing framework time out and nukes our test including log gathering.

I've also used gsutil download all those pod logs from the last CI run.  I'll search that data this weekend (assuming we don't lose power for a while from the hurricane :-) ), and see if I see the same "stuck accessing the cache in the pipelinerun messages.

And of course, I'm not sure how moving from the persistent volume to the ephemeral volume has any bearing.  Any chance the localstack stuff was adding some sort of stability buffer?

Of course I can switch this PR back to PVs if need be (I've done that before).  And we can d a test PR off of main branch.

But I'm beginning to consider trying a branch that reverts the localstack removal, and see if that matters.

gabemontero commented:
so I couldnt' wait, and yep,  from the gsutil download, and ton of 

`jvm-build-service-test-namespace-f7lzc_e8e5c8a844d3d255429e9f55730b1e8d-build-0-task-pod_step-build.log:[INFO] Retrying request to {}->http://jvm-build-workspace-artifact-cache.jvm-build-service-test-namespace-f7lzc.svc.cluster.local:80
`

@stuartwdouglas that seems to be what we need to zero in on

stuartwdouglas commented:
Yea, it is clearly becoming unresponsive. I don't think it is localstack related though. It would be really useful to get some thread dumps to see what it is doing.

My guess is the worker threads are getting stuck, and the health check is just running in the IO thread so it still reports the pod being health. 

I have actually seen a deadlock before related to the Apache HTTP client, so I will investigate further if I get time over the weekend.

gabemontero commented:
> Yea, it is clearly becoming unresponsive. I don't think it is localstack related though. It would be really useful to get some thread dumps to see what it is doing.
> 
> My guess is the worker threads are getting stuck, and the health check is just running in the IO thread so it still reports the pod being health.
> 
> I have actually seen a deadlock before related to the Apache HTTP client, so I will investigate further if I get time over the weekend.

Next time I run locally, assuming this is repeatable, I'll see about performig an `oc rsh ...` into of the underlying pods and running `kill -3 pid` or `jstack pid` (if jstack is available in that image)

stuartwdouglas commented on pkg/reconciler/userconfig/userconfig.go:
```diff
@@ -216,6 +209,10 @@ func (r *ReconcilerUserConfig) cacheDeployment(ctx context.Context, log logr.Log
 	create := false
 	if err != nil {
 		if errors.IsNotFound(err) {
+			qty, err := resource.ParseQuantity(settingOrDefault(userConfig.Spec.Storage, v1alpha1.ConfigArtifactCacheStorageDefault))
```
Silly question, but I can't see to see this actually being set in the tests?

stuartwdouglas replied:
Oh, it defaults to 10gb, never mind. I have not had enough coffee yet.

gabemontero commented:
confirmed main brach independent of this PR is hitting the hang: https://github.com/redhat-appstudio/jvm-build-service/pull/279#issuecomment-1264339562

gabemontero commented:
ok I believe I have changes to the periodic test that
- fix the timeout / did not complete in time / but still gather data vs. have the overall test timeout nuke us and prevent our debug data capture from working
- updated the report to highlight the number of DB's still in building state
- added a toggle button to allow for filtering of just the DB's in building state 

I'm making one more pass locally @stuartwdouglas to validate the report tweaks (I have a couple of minor tweaks since my first run today that minimally handles the first bullet) and then I'll push here.

At that point, I'm thinking of merging this given 
a) I do believe I've sorted out the default mem needed for the cache with ephemeral volumes (via my `oc adm top` examinations)
b) we know the cache hang is unrelated to that switch in the volume now
c) my test fixes will help your PRs for diagnosing the cache hang
b) merging this will allow progress on adding jvm-build-service back to infra-deployments (though I suspect at least one more tweak to the kcp yaml here will be needed)

gabemontero commented on openshift-with-appstudio-test/e2e/periodic_test.go:
```diff
@@ -170,6 +169,7 @@ func TestServiceRegistry(t *testing.T) {
 		state := sync.Map{}
 		var changed uint32
 		exitForLoop := false
+		timeoutChannel := time.After(90*time.Minute)
```
@stuartwdouglas - I did bump the timeout down from 2 hours to 90 minutes (where the overall test is aborted by the framework after 3 hours).

I did this in part to have the test fail a bit sooner while the artifact cache hang is being debugged, but to let it run long enough to make sure the hang manifests.

Once the hang has been fixed, I certainly could see changing this back to 2 hours (I think we need a little buffer for cluster bringup and the initial build) if we get back to a point where things are fine but we determine we just need more time.

gabemontero commented:
/test jvm-build-service-in-repo-apicurio-e2e

gabemontero commented:
OK the test changes look "good"
- the timeout processing worked correctly
`periodic_test.go:344: CHECK THE GENERATED REPORT AND SEE IF THE DEPENDENCYBUILDS IN BUILDING STATE ARE STUCK OR JUST SLOW `
- in the report, you see 50 active builds at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1576261533959196672/artifacts/jvm-build-service-in-repo-apicurio-e2e/jvm-build-service-in-repo-apicurio-e2e/artifacts/jvm-build-service-report/index.html
- which is expected since we do 50 pipelineruns at a time with throttling
- hitting toggle buttons so just DB's in Building state are visible, it did not take me long to find a build stuck trying to download a dependency, but the artifact cache is hung, such as https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/redhat-appstudio_jvm-build-service/270/pull-ci-redhat-appstudio-jvm-build-service-main-jvm-build-service-in-repo-apicurio-e2e/1576261533959196672/artifacts/jvm-build-service-in-repo-apicurio-e2e/jvm-build-service-in-repo-apicurio-e2e/artifacts/jvm-build-service-report/DependencyBuildStateBuilding/080dbba8b3ffba35739ebe5bce69a2be080dbba8b3ffba35739ebe5bce69a2be-build-0-task-podstep-build

I'll wait to see if you have time during you Sunday @stuartwdouglas to take a peek, agree with the changes, and merge.  But at the latest, I want to merge this first thing Monday my time to pivot back to the infra-deployment changes.

I do think these changes will help with debug around the cache hang.
